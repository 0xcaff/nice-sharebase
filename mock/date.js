// Sets any calls to date to return a constant time with a lot of zeros.
const constantDate = new Date(1500000000000);

global.Date.now = () => constantDate;