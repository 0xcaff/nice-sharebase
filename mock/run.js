require('babel-register');
require("regenerator-runtime/runtime");

const app = require('./server');
app.listen(30101, '0.0.0.0', function() { console.log('started mock server on port 30101') });
