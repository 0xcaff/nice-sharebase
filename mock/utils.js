import fs from 'fs';
import url, { URLSearchParams } from 'url';
import path from 'path';

// Returns the result of the first value in elems send through predicate which didn't throw an error.
export async function first(elems, predicate) {
	for (let elem of elems) {
		const { result, error } = await tryIt(_ => predicate(elem));
		if (!error) {
			return result;
		}
	}
}

// Wraps fs.readFile to use promises.
export const readFile = (...args) =>
  new Promise((resolve, reject) =>
    fs.readFile(...args, (err, data) => {
      if (err) { reject(err) } else { resolve(data) }
    })
  );

// Async returns an object with an error if one was caught. Otherwise returns the
// result.
export const tryIt = async (it) => {
  try {
    return {result: await it()};
  } catch(e) {
    return {error: e};
  }
};

// If a folder is requested with embeddings added, add we add them here.
export async function embedIfNeeded(folder, location) {
  const params = new URLSearchParams(location.search);
  const embed = params.get('embed');

  if (!embed) {
    return folder;
  }

  const embeds = embed.split(',');
  folder = JSON.parse(folder);

  const id = folder['FolderId'];
  if (embeds.includes('f')) {
    // try to embed folders
    const file = path.resolve(__dirname, `api/folders/${id}/folders.json`);

    const { result, error } = await tryIt(_ => readFile(file));
    if (!error && result) {
      const folders = JSON.parse(result);
      folder["Embedded"]["Folders"].push(...folders);
    }
  }

  // try to get folders
  if (embeds.includes('d')) {
    // try to embed documents
    const file = path.resolve(__dirname, `api/folders/${id}/documents.json`);

    const { result, error } = await tryIt(_ => readFile(file));
    if (!error && result) {
      const documents = JSON.parse(result);
      folder["Embedded"]["Documents"].push(...documents);
    }
  }

  return JSON.stringify(folder);
};

export async function handle(p) {
  const location = url.parse(p);
  p = location.pathname.slice(1);

	const paths = [`${p}.json`, `${p}/index.json`]
 		.map(raw => path.resolve(__dirname, raw));

  const result = await first(paths, path => readFile(path));
  if (result !== undefined) {
    const status = 200;
    const message = await embedIfNeeded(result, location);
    const headers = {'Content-Type': 'application/json'};

    return { status, message, headers };
  } else {
    const status = 404;
		const pre = `Path not found for ${p}. Tried:`;
		const tried = paths.map(path => '\t' + path).join('\n');
    const message = pre + '\n' + tried;

    return { status, message };
  }
}
