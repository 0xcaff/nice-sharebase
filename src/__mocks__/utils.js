const utils = require.requireActual('../utils');

utils.token = jest.fn(() => {
  const calls = utils.token.mock.calls.length;
  return `Session${calls}`;
});

module.exports = utils;
