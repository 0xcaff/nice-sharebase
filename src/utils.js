import crypto from 'crypto';

const get = (prop) => (obj) => obj[prop];

// Creates a new object copying all functions and replacing other values with
// getter functions.
export const getters = (kvs) =>
  Object.entries(kvs)
    .reduce(
      (acc, [key, value]) => {
        if (value instanceof Function) {
          // just copy it
          acc[key] = value;
        } else {
          acc[key] = get(value);
        }
        return acc;
      },
      {},
    );

// Returns a base64 encoded string of the string passed in.
export const base64Encode = (toEncode) => Buffer.from(toEncode).toString('base64');
export const base64Decode = (encoded) => Buffer.from(encoded, 'base64').toString();

// A function which returns its first argument.
export const nop = (arg) => arg;

// Copys the properties and their values of the input object to a new output
// object and returns it.
export const filter = (obj) =>
  Object.entries(obj)
    .reduce(
      (acc, [k, v]) => { if (v) { acc[k] = v } return acc; },
      {},
    );

// Creates a cryptographically random sequence of characters.
export const token = async () => base64Encode(await rand(256));

// Wraps the crypto.randomBytes API to use promises.
const rand = (size) =>
  new Promise((resolve, reject) =>
    crypto.randomBytes(size, (err, buf) => {
      if (err) { reject(err) }
      resolve(buf);
    })
  );

// A functional way to throw errors. This is useful for throwing errors when
// destructuring an object when a value isn't provided.
export const toss = (err) => { throw err };
export const required = (what, foor) =>
  toss(new TypeError(`${what} is a required parameter to ${foor}`));

export const MS = 1;
export const SECONDS = 1000 * MS;
export const MINUTES = 60 * SECONDS;
export const HOURS = 60 * MINUTES;
