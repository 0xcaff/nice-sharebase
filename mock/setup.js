// Sets up nock to respond to requests to the mock endpoint.
import nock from 'nock';

import { handle } from './utils';

export const ENDPOINT = 'http://mock.com/';

export const logs = [];

global.beforeEach && beforeEach(() => logs.length = 0);

const setup = app => app
  .get(() => true)
  .reply(function(url, body, cb) {
    h(url)
      .then((...args) => {
        // setup for next request
        setup(app);
        cb(null, ...args);
      })
      .catch(err => cb(err));
  });

const app = nock(ENDPOINT);
setup(app);

async function h(path) {
  log(path);
  const { status, message, headers } = await handle(path);
  return [status, message, headers];
}

const log = (path) => logs.push(path);

