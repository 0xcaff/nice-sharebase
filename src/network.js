import url from 'url';
import 'isomorphic-fetch';

import { throwOnFail } from './errors';
import { base64Encode, MINUTES, required, token } from './utils';

export const initNetwork = ({ base, transform, logs, session }) => {
  const network = {};
  network.req = req({ network, base, transform, logs });
  network.renew = renew({ network, session });
  network.request = request({ network, session });
  return network;
};

// Makes a request to the original ShareBase API getting new tokens if needed.
const req = ({ base, transform, logs }) =>
  async (path, headers) => {
    const fullPath = url.resolve(base, path);
    logs.push(fullPath);
    const req = transform(new Request(fullPath, { headers }));
    const resp = await throwOnFail(await fetch(req));
    const body = await resp.json();

    return body;
  };

// Renews or creates the session using the information provided. It updates
// context.session, context.sessionStore and returns the response received from
// the server along with the token to use to access the just renewed session.
const renew = ({ network, session }) =>
  async (
    email = required('email', 'renew'),
    password = required('password', 'renew')
  ) => {
    const { store } = session;

    const inner = base64Encode(`${email}:${password}`);
    const headers = new Headers({'Authorization': `Basic ${inner}`});
    const authed = await network.req('authenticate', headers);

    // update session
    if (!session.me) {
      // this is a new session, update the session store and context
      const delegate = await token();

      const me = { email, password };
      authed.delegate = delegate;
      store.set(delegate, me);

      session.me = me;
    }

    const { me } = session;
    me.expires = +Date.parse(authed['ExpirationDate']);
    me.authed = authed;

    // expose the delegate token
    return authed;
  };

const request = (context) =>
  async (path) => {
    const { session, network: { renew, req } } = context;
    const { me = required('me', 'req') } = session;

    // handle renewals
    const timeToExpire = me.expires - Date.now();
    if (!me.renewalPromise && timeToExpire < 5 * MINUTES) {
      // renew the credentials
      const { email, password } = me;

      me.renewalPromise = renew(email, password);
    }

    if (me.renewalPromise) {
      // if the renewal fails, the following line will throw and any request
      // which triggered a renewal will fail
      await me.renewalPromise;
      me.renewalPromise = null;
    }

    // we have an active session, use it
    const token = me.authed['Token'];
    const headers = new Headers({ 'Authorization': `PHOENIX-TOKEN ${token}` });
    return await req(path, headers);
  };