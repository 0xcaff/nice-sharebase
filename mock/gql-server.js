import gql from 'express-graphql';
import express from 'express';
import url from 'url';
import { schema, createContext } from '../index.js';

import { ENDPOINT, logs } from './setup';

const app = express();

app.use('/graphql', gql(
  async (req, res, params) => {
    const base = url.resolve(ENDPOINT, 'api/');
    const context = createContext({ base });

    return {
      schema,
      context,
      extensions: (document, variables, operationName, result) => {
        const l = logs.slice();
        logs.length = 0;
        return l;
      },
      graphiql: true,
      pretty: true,
    };
  }
));

app.listen(3010, '0.0.0.0', _ => console.log('listening on 0.0.0.0:3010'));
