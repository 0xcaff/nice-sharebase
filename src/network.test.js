jest.mock('./utils');
import '../mock/date';

import url from 'url';
import { DEFAULT_ENDPOINT, setup } from '../mock/setup';
import { nop } from './utils';
import { initNetwork } from './network';

const base = url.resolve(DEFAULT_ENDPOINT, 'api/');
const transform = nop;

const allowed = new Set();
setup({ allowed });

beforeEach(() => {
  allowed.clear();
});

it('should authenticate with the backend', async () => {
  const session = { store: new Map() };
  const logs = [];

  const network = initNetwork({ base, transform, logs, session });
  const resp = await network.renew('sam.babic@onbase.com', 'password');

  // the sesion should be updated so future requests don't reauth
  expect(session.me).toBeTruthy();

  expect(resp).toMatchSnapshot();
  expect(session).toMatchSnapshot();
  expect(logs).toMatchSnapshot();
});

it('should make requests using a phoenix token', async () => {
  const token = 'PHNX-BACKEND-TOKEN';
  const session = { store: new Map(), me: null, token };
  const logs = [];

  allowed.add(token);
  const network = initNetwork({ base, transform, logs, session });
  const resp = await network.request('libraries');

  expect(resp).toMatchSnapshot();
  expect(session).toMatchSnapshot();
  expect(logs).toMatchSnapshot();
});

it('should get pnx tokens without updating context', async () => {
  const token = 'NICE-SHAREBASE-ACCESS-TOKEN';
  const email = 'sam.babic@onbase.com';
  const password = 'password';

  const logs = [];
  const me = { email, password };
  const store = new Map();
  store.set(token, me);

  const session = { store, me, token };
  const network = initNetwork({ base, transform, logs, session });

  const pnxToken = await network.getPnxToken(email, password);

  expect(pnxToken).toMatchSnapshot();
  expect(session).toMatchSnapshot();
  expect(logs).toMatchSnapshot();
});
