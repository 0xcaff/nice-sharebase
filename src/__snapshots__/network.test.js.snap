// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should authenticate with the backend 1`] = `
Object {
  "ExpirationDate": "2017-07-14T02:50:00.000Z",
  "Token": "Session1",
  "UserId": 0,
  "UserName": "sam.babic@onbase.com",
  "delegate": "Session2",
}
`;

exports[`should authenticate with the backend 2`] = `
Object {
  "me": Object {
    "authed": Object {
      "ExpirationDate": "2017-07-14T02:50:00.000Z",
      "Token": "Session1",
      "UserId": 0,
      "UserName": "sam.babic@onbase.com",
      "delegate": "Session2",
    },
    "email": "sam.babic@onbase.com",
    "expires": 1500000600000,
    "password": "password",
  },
  "store": Map {
    "Session2" => Object {
      "authed": Object {
        "ExpirationDate": "2017-07-14T02:50:00.000Z",
        "Token": "Session1",
        "UserId": 0,
        "UserName": "sam.babic@onbase.com",
        "delegate": "Session2",
      },
      "email": "sam.babic@onbase.com",
      "expires": 1500000600000,
      "password": "password",
    },
  },
}
`;

exports[`should authenticate with the backend 3`] = `
Array [
  "http://mock.com/api/authenticate",
]
`;

exports[`should get pnx tokens without updating context 1`] = `
Object {
  "ExpirationDate": "2017-07-14T02:50:00.000Z",
  "Token": "Session3",
  "UserId": 1,
  "UserName": "sam.babic@onbase.com",
}
`;

exports[`should get pnx tokens without updating context 2`] = `
Object {
  "me": Object {
    "email": "sam.babic@onbase.com",
    "password": "password",
  },
  "store": Map {
    "NICE-SHAREBASE-ACCESS-TOKEN" => Object {
      "email": "sam.babic@onbase.com",
      "password": "password",
    },
  },
  "token": "NICE-SHAREBASE-ACCESS-TOKEN",
}
`;

exports[`should get pnx tokens without updating context 3`] = `
Array [
  "http://mock.com/api/authenticate",
]
`;

exports[`should make requests using a phoenix token 1`] = `
Array [
  Object {
    "IsPrivate": true,
    "LibraryId": 406,
    "LibraryName": "My Library",
    "Links": Object {
      "Folders": "https://app.sharebase.com/ShareBaseAPI/api/libraries/406/folders",
      "Self": "https://app.sharebase.com/ShareBaseAPI/api/libraries/406",
    },
  },
  Object {
    "IsPrivate": false,
    "LibraryId": 1782,
    "LibraryName": "Iron Man Suit",
    "Links": Object {
      "Folders": "https://app.sharebase.com/ShareBaseAPI/api/libraries/1782/folders",
      "Self": "https://app.sharebase.com/ShareBaseAPI/api/libraries/1782",
    },
  },
]
`;

exports[`should make requests using a phoenix token 2`] = `
Object {
  "me": null,
  "store": Map {},
  "token": "PHNX-BACKEND-TOKEN",
}
`;

exports[`should make requests using a phoenix token 3`] = `
Array [
  "http://mock.com/api/libraries",
]
`;
